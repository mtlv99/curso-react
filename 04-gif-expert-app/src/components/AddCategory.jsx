import { useState } from 'react';

// export default function AddCategory({ setCategory }) {
export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();

    const cleanValue = inputValue.trim();

    // el input debe tener al menos 2 caracteres para poder ser agregado al state.
    if (cleanValue.length <= 1) return;

    // Limpiar state del input luego de agregar el valor
    setInputValue('');

    // Emitir valor guardado al padre
    onNewCategory(cleanValue);
  };

  /*
  // Sin invertir la responsabilidad: el setState se hace acá.
  const onSubmit = (event) => {
    event.preventDefault();

    // el input debe tener al menos 2 caracteres para poder ser agregado al state.
    if (inputValue.trim().length <= 1) return;

    // NO es necesario pasar el state de 'category' por props! Solo usar el callback de setCategory.
    setCategory((categories) => [inputValue, ...categories]);

    // Limpiar state del input luego de agregar el valor
    setInputValue('');
  };
  */

  /*
    "What's the name of both of these techniques? (React hooks)"
    1. onChange={onInputChange} ---------------------> passing the handler function as a prop
    2. onChange={(event) => onInputChange(event)} ---> using an arrow function
    Nota: Se puede mejorar usando useCallback para que este memoized
  */
  return (
    <form onSubmit={onSubmit}>
      {/* Recordar que para los input se deben vincular tanto el value, como el onChange */}
      <input
        type="text"
        placeholder="Escribe algo interesante (café, galletas, navidad)..."
        value={inputValue}
        onChange={onInputChange}
        // Lo de arriba equivale a:
        // onChange={(event) => onInputChange(event)}
      />
    </form>
  );
};
