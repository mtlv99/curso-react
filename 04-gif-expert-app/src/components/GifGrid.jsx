
import useFetchGifs from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>

      {isLoading && (<h2>Cargando...</h2>)}

      <div className="card-grid">
        {/* TambiÃ©n se puede hacer prop spreading, aunque el linter dice que es mala practica... */}
        {/* {images.map((image) => <GifGridItem key={image.id} {...image} />)} */}
        {images.map(({ title, url, id }) => <GifGridItem key={id} title={title} url={url} id={id} />)}
      </div>

    </>
  );
};
