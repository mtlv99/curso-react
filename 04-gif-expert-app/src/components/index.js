// ARCHIVO DE BARRIL / ARCHIVO DE INDEX/INDICE
export * from './AddCategory';
export * from './GifGrid';
export * from './GifGridItem';
