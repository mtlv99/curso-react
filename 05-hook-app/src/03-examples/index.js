export * from './LoadingQuote';
export * from './Quote';
// Este puede crear un ciclo de dependencias (dependency cycle), descomentar para poder ver
// la regla de eslint mostrando un error sobre ello.
// export * from './MultipleCustomHooks';
