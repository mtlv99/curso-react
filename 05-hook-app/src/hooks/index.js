// ARCHIVO DE BARRIL / ARCHIVO DE INDEX/INDICE
export * from './useCounter';
export * from './useFetch';
export * from './useForm';
