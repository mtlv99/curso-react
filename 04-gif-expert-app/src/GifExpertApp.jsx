import React, { useState } from 'react';

// Los archivos de index/indice o de barril son utiles para este tipo de importaciones,
// 2 componentes en un mismo import
import { AddCategory, GifGrid } from './components';

const GifExpertApp = () => {
  const [category, setCategory] = useState([]);

  const onAddCategory = (newCategory) => {
    // Valida que no exista una categoria con el mismo nombre, para evitar el error de las keys en los map.
    // No es mejor usar un uuid?
    if (category.includes(newCategory)) return;

    // Nota:
    // (A partir de React 18, si hay varios setState seguidos, React esperará a que ésta funcion principal
    // terminar para hacer el re-renderizado. En versiones anteriores, se hace un rerender por cada setState )

    setCategory([newCategory, ...category]);
    // setCategory((categories) => [newCategory, ...categories]);
    // Los llamados de arriba son equivalentes.
    // Aunque en componentes hijos, NO es necesario pasar el state de 'category' por props! Solo usar el callback de setCategory.
  };

  return (
    <>
      <h1>Buscador de GIF&apos;s</h1>

      <AddCategory
        // category={category}
        // setCategory={setCategory}
        // Recordar que cualquier funcion que tenga un 'on' al principio,
        // indica que emite un valor hacia afuera del componente (usualmente).
        onNewCategory={onAddCategory}
      />

      {category.map((categoryItem) => (
        <GifGrid
          key={categoryItem}
          category={categoryItem}
        />
      ))}
    </>


  );
};

export default GifExpertApp;
