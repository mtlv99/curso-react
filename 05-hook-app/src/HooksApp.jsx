import { CounterApp } from './01-useState/CounterApp';
import { CounterWithCustomHook } from './01-useState/CounterWithCustomHook';
import { FormWithCustomHook } from './02-useEffect/FormWithCustomHook';
import { SimpleForm } from './02-useEffect/SimpleForm';
import { MultipleCustomHooks } from './03-examples/MultipleCustomHooks';
import { FocusScreen } from './04-useRef/FocusScreen';
import { Layout } from './05-useLayoutEffect/Layout';
import { LayoutComments } from './05-useLayoutEffect/LayoutComments';
import { Memorize } from './06-memos/Memorize';
import { MemoHook } from './06-memos/MemoHook';

export const HooksApp = () => (
  // <CounterApp />
  // <CounterWithCustomHook />
  // <SimpleForm />
  // <FormWithCustomHook />
  // <MultipleCustomHooks />
  // <FocusScreen />
  // <Layout />
  // <LayoutComments />
  // <Memorize />
  <MemoHook />
);
