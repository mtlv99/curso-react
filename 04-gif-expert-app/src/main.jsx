import React from 'react';
import ReactDOM from 'react-dom/client';
import GifExpertApp from './GifExpertApp';
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  // React renderizará 2 veces los componentes cuando se usa el Strict Mode,
  // esto para asegurarse de que no existan loops y otras malas practicas que serán
  // detectadas en el momento.
  <React.StrictMode>
    <GifExpertApp />
  </React.StrictMode>,
);
