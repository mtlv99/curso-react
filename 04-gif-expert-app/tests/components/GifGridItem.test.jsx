import { render } from '@testing-library/react';
import { GifGridItem } from '../../src/components';


describe('Pruebas en <GifGridItem/>', () => {
  const title = 'Cat';
  const url = 'https://placekitten.com/400/300';

  test('Debe de hacer match con el snapshot', () => {
    const { container } = render(<GifGridItem title={title} url={url} />);
    expect(container).toMatchSnapshot();
  });
});
