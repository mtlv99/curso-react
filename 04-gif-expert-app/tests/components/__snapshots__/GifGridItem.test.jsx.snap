// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem/> Debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Cat"
      src="https://placekitten.com/400/300"
    />
    <p>
      Cat
    </p>
  </div>
</div>
`;
